### PR TITLE
Closes #5998: Emit Facts for WebExtension Telemetry

### LINKED ISSUE #6174: Introduce a Cache for TopSites to improvement Fenix Cold Start performance
## Description of Problem

Fenix measures its cold start performance (at the moment) when `onBindViewHolder` occurs for its RecyclerView TopSites. On a G5+ we saw a cold start up time of ~870ms.  I noticed that we had multiple layout passes triggered and a wait time from our `homeFragmentStore.dispatch` which lead to a noticeable delay

## Proposed Solution

Introduce a cache in the form of SharedPreferences where we can load the Users' TopSites through there on start.  I.e:

```
 fun getTopSitesFromPrefs() : List<TopSite>
    {
        val topsites = mutableListOf<TopSite>()
        pref.all.forEach {
            topsites.add(
                TopSitePrefs(
                    Entity(
                        it.key, it.value.toString(), UUID.randomUUID().leastSignificantBits
                    )
                )
            )
        }
        return topsites
    }
```

where `Entity()` is the same as our `TopSiteEntity` at the moment.

This solution seems to bring down start up time on the G5+ to 700ms.
Here are two apks built with TopSites from Room and TopSites from SharedPref: https://drive.google.com/open?id=1eMtvsLVw52zH7XA7YzUE1sIcxmtRUYWm

### PR DESCRIPTION
**Pull request by gabrielluong**
Fixes #5998 

---

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by gabrielluong** [comment]
In Fenix, we intend on sending a metric pings indicating which addons are installed and enabled. We emit the list of installed and enabled addons as facts from A-C when the web extensions are finished initializing.

File: components/support/webextensions/README.md [hunk 1]
```diff
@@ -14,6 +14,22 @@ Use Gradle to download the library from [maven.mozilla.org](https://maven.mozill
...
 ```
 
+## Facts
+
+This component emits the following [Facts](../../support/base/README.md#Facts):
+
+| Action      | Item                      | Extras            | Description                  |
+|-------------|---------------------------|-------------------|------------------------------|
+| Interaction | web_extension_initialized | `extensionExtras` | Web extension is initalized. |
+
```
Comment (Amejia481) [inline components/support/webextensions/README.md]: nit: should this be `web_extensions_initialized`?

File: components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt [hunk 2]
```diff
@@ -219,6 +220,7 @@ object WebExtensionSupport {
         runtime.listInstalledWebExtensions(
             onSuccess = {
                 extensions -> extensions.forEach { registerInstalledExtension(store, it) }
+                emitWebExtensionInitializedFact(extensions)
```
Comment (Amejia481) [inline components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt]: nit: `emitWebExtensionsInitializedFact`?

**Comment by Amejia481** [comment]
@gabrielluong I think we could add a test to validate facts are getting collected, we can added [here](https://github.com/mozilla-mobile/android-components/blob/80bf354477f53df75fb3278693e01788f15cd999/components/support/webextensions/src/test/java/mozilla/components/support/webextensions/WebExtensionSupportTest.kt#L90) after `WebExtensionSupport.initialize` gets called, similar to what you did [here](https://github.com/mozilla-mobile/android-components/pull/5010/files#diff-1679616fca013a0fac84f16e5b886549R132). What do you think?

File: components/support/webextensions/README.md (hunk 1 above)
Comment (gabrielluong) [reply components/support/webextensions/README.md]: Fixed.

File: components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt (hunk 2 above)
Comment (gabrielluong) [reply components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt]: Fixed.

File: components/support/webextensions/src/main/java/mozilla/components/support/webextensions/facts/WebExtensionFacts.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    Fact(
+        Component.FEATURE_ADDONS,
+        action,
+        item,
+        value,
+        metadata
+    ).collect()
+}
+
+internal fun emitWebExtensionInitializedFact(extensions: List<WebExtension>) {
```
Comment (csadilek) [inline components/support/webextensions/src/main/java/mozilla/components/support/webextensions/facts/WebExtensionFacts.kt]: nit: this should be plural? `emitWebExtenionsInitializedFact`.

File: components/support/webextensions/src/main/java/mozilla/components/support/webextensions/facts/WebExtensionFacts.kt [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    val installedAddons = extensions.filter { !it.isBuiltIn() }
+    val enabledAddons = installedAddons.filter { it.isEnabled() }
+    emitWebExtensionFact(
+        Action.INTERACTION,
+        WebExtensionFacts.Items.WEB_EXTENSION_INITIALIZED,
+        metadata = mapOf(
```
Comment (csadilek) [inline components/support/webextensions/src/main/java/mozilla/components/support/webextensions/facts/WebExtensionFacts.kt]: I assume all installed and all enabled ones are good enough for now? We don't want to report the "unsupported" ones? You can get those by checking the disabled flags for `APP` if you need to.

File: components/support/webextensions/src/main/java/mozilla/components/support/webextensions/facts/WebExtensionFacts.kt [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            "installed" to installedAddons,
```
Comment (csadilek) [inline components/support/webextensions/src/main/java/mozilla/components/support/webextensions/facts/WebExtensionFacts.kt]: You're emitting the whole extension object as part of the Fact. I think that's fine but, wondering if the ID would be good enough?

File: components/support/webextensions/README.md [hunk 6]
```diff
@@ -14,6 +14,22 @@ Use Gradle to download the library from [maven.mozilla.org](https://maven.mozill
 implementation "org.mozilla.components:support-webextensions:{latest-version}"
 ```
 
+## Facts
+
+This component emits the following [Facts](../../support/base/README.md#Facts):
+
+| Action      | Item                      | Extras            | Description                  |
+|-------------|---------------------------|-------------------|------------------------------|
+| Interaction | web_extension_initialized | `extensionExtras` | Web extension is initalized. |
```
Comment (csadilek) [inline components/support/webextensions/README.md]: nit: web_extensions_initialized?

File: components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt (hunk 2 above)
Comment (csadilek) [inline components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt]: A test would be good to verify that the fact is emitted on successful init. :)

File: components/support/webextensions/src/main/java/mozilla/components/support/webextensions/facts/WebExtensionFacts.kt (hunk 4 above)
Comment (gabrielluong) [reply components/support/webextensions/src/main/java/mozilla/components/support/webextensions/facts/WebExtensionFacts.kt]: Yea, just need installed and enabled for now.

**Pull request merged**